File: frontend/src/types/knowledge.ts
/**
 * 知识库模型 —— 前端 mock 版。
 * KnowledgeBase 是一个逻辑分组(比如"临床指南"、"内部 SOP"),
 * 里面装若干 KnowledgeDoc(纯文本 / Markdown 段落 / 本地上传文件)。
 * search_kb 工具会在启用的知识库范围内做关键字命中,把片段回填到工具轨迹。
 */
export type KbDocType = 'markdown' | 'text' | 'link' | 'upload';

export interface KnowledgeDoc {
  id: string;
  title: string;
  /** 全文内容(markdown/text/upload 解析后的纯文本)或外链摘要 */
  content: string;
  type: KbDocType;
  /** 逗号分隔的标签,便于筛选 */
  tags: string[];
  /** 若 type==='link' 才有 */
  url?: string;
  /** 'manual' = 文本编辑器创建;'upload' = 本地文件上传 */
  source?: 'manual' | 'upload' | '';
  /** upload 才有: 原始文件名 */
  original_name?: string;
  /** upload 才有: 原始 Content-Type */
  mime_type?: string;
  /** upload 才有: 原始字节数 */
  size_bytes?: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeBase {
  id: string;
  name: string;
  description: string;
  /** UI 展示色 */
  color: string;
  builtin?: boolean;
  /** 单 KB 上传配额(字节),后端默认 1 GiB */
  quota_bytes?: number;
  docs: KnowledgeDoc[];
  created_at: string;
  updated_at: string;
}

/** 检索命中片段 */
export interface KbSearchHit {
  kb_id: string;
  kb_name: string;
  doc_id: string;
  doc_title: string;
  /** 命中的 chunk 唯一 id("doc-xxx:N")。 */
  chunk_id?: string;
  /** 命中处上下文片段(chunk 原文)。 */
  snippet: string;
  /** chunk 位置: "第 3 页" / "## 心衰诊治" 等(阶段 5 续 3 新增)。 */
  location?: string;
  score: number;
  /** 命中文档的原始来源链接(来自 KnowledgeDoc.url),供引用可点击跳转 */
  url?: string;
}

/** 阶段 5 续 5: 文档结构化抽取结果(后端 sidecar 返)。 */
export interface DocStructureBlock {
  type: 'title' | 'text' | 'list' | 'table' | 'formula' | 'figure' | string;
  content?: string;
  latex?: string;
  caption?: string;
}

export interface DocStructurePage {
  page: number;
  blocks: DocStructureBlock[];
}

export interface DocStructure {
  doc_id: string;
  kb_id: string;
  pages_n: number;
  markdown: string;
  /** VLM 原始返的 pages/blocks JSON, 后端会做 json.Unmarshal。 pymupdf 时是 page_chunks 数组而非 {pages, blocks}, 需用 structure_schema 分支 */
  structure: { pages?: DocStructurePage[]; merged_tables?: DocStructureTable[] };
  /** P91: 后端填的 schema 标识 "qwen-vl" | "pymupdf" | "" (空 = 旧数据 Qwen-VL) */
  structure_schema?: 'qwen-vl' | 'pymupdf' | string;
  /** P85: pdfcpu 抽出的 embedded images, [{idx, filename, file_type, byte_size, width, height, page_nr, obj_nr}] */
  images: DocStructureImage[];
  updated_at: string;
}

export interface DocStructureTable {
  header: string[];
  rows: string[][];
  source_pages: number[];
}

/** P85: 抽出的图片元数据 — 二进制走 GET /api/kb/{id}/docs/{docId}/images/{filename}。 */
export interface DocStructureImage {
  idx: number;
  /** 相对路径文件名, 如 "image_001.jpg"。 前端拼 URL 走 handleGetDocImage。 */
  filename: string;
  file_type: string; // jpg / png / jpeg / webp / tif
  byte_size: number;
  width: number;
  height: number;
  page_nr: number; // 1-based
  obj_nr: number;
}

File: backend/internal/domain/knowledge.go
package domain

// KbDocType 文档类型 —— 与前端 KbDocType 对齐。
type KbDocType string

const (
	KbDocTypeMarkdown KbDocType = "markdown"
	KbDocTypeText     KbDocType = "text"
	KbDocTypeLink     KbDocType = "link"
	// KbDocTypeUpload 来自本地文件上传的文档 —— 原始二进制存 kb_doc_uploads
	// 表,Content 字段是解析后的纯文本,供 search_kb / enrich_content 引用。
	KbDocTypeUpload KbDocType = "upload"
)

// KbDefaultQuotaBytes 每个 KB 的默认上传配额(1 GiB)。
// 上传 handler 会在请求前检查 used + fileSize <= kb.QuotaBytes,
// 超出返 413。前端按此值显示进度条。
const KbDefaultQuotaBytes int64 = 1 << 30

// KnowledgeDoc 一篇文档。
// 对齐 frontend/src/types/knowledge.ts KnowledgeDoc。
type KnowledgeDoc struct {
	ID        string    `json:"id"`
	Title     string    `json:"title"`
	Content   string    `json:"content"`
	Type      KbDocType `json:"type"`
	Tags      []string  `json:"tags"`
	URL       string    `json:"url,omitempty"`
	CreatedAt string    `json:"created_at"`
	UpdatedAt string    `json:"updated_at"`

	// 上传文档的元数据(Source == "upload" 才有意义)。
	// Source 为空或 "manual" 时均为文本编辑器创建,无原始二进制。
	Source       string `json:"source,omitempty"`        // "manual" | "upload" | ""
	OriginalName string `json:"original_name,omitempty"` // 上传时的原始文件名
	MimeType     string `json:"mime_type,omitempty"`     // 原始 Content-Type
	SizeBytes    int64  `json:"size_bytes,omitempty"`    // 原始字节数
}

// KnowledgeBase 知识库(逻辑分组)。
// 对齐 frontend/src/types/knowledge.ts KnowledgeBase。
type KnowledgeBase struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	Description string          `json:"description"`
	Color       string          `json:"color"`
	Builtin     bool            `json:"builtin,omitempty"`
	QuotaBytes  int64           `json:"quota_bytes,omitempty"` // 单 KB 上传配额,默认 KbDefaultQuotaBytes
	Docs        []*KnowledgeDoc `json:"docs"`
	CreatedAt   string          `json:"created_at"`
	UpdatedAt   string          `json:"updated_at"`
}

func (k *KnowledgeBase) GetID() string   { return k.ID }
func (k *KnowledgeBase) IsBuiltin() bool { return k.Builtin }
func (k *KnowledgeBase) SetID(id string) { k.ID = id }

// Normalize 保证切片非 nil,补默认颜色。
func (k *KnowledgeBase) Normalize() {
	if k.Docs == nil {
		k.Docs = []*KnowledgeDoc{}
	}
	for _, d := range k.Docs {
		if d == nil {
			continue
		}
		if d.Tags == nil {
			d.Tags = []string{}
		}
		if d.Type == "" {
			d.Type = KbDocTypeMarkdown
		}
	}
	if k.Color == "" {
		k.Color = "#2b57d6"
	}
	// 老数据 / 老 fixture 走 Normalize 时也补默认配额,让 quota 字段始终有值。
	if k.QuotaBytes <= 0 {
		k.QuotaBytes = KbDefaultQuotaBytes
	}
}

// KbSearchHit 检索命中片段。
// 对齐 frontend/src/types/knowledge.ts KbSearchHit。
type KbSearchHit struct {
	KBID     string  `json:"kb_id"`
	KBName   string  `json:"kb_name"`
	DocID    string  `json:"doc_id"`
	DocTitle string  `json:"doc_title"`
	Snippet  string  `json:"snippet"`
	Score    float64 `json:"score"`
	// URL 命中文档的原始来源链接(来自 KnowledgeDoc.URL),供引用可点击跳转;无则为空。
	URL string `json:"url,omitempty"`
}
